lib/avalon/variations_timeline_importer.rb

```diff
@@ -0,0 +1,239 @@
+# frozen_string_literal: true
+
+# Copyright 2011-2018, The Trustees of Indiana University and Northwestern
+#   University.  Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software distributed
+#   under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+#   CONDITIONS OF ANY KIND, either express or implied. See the License for the
+#   specific language governing permissions and limitations under the License.
+# ---  END LICENSE_HEADER BLOCK  ---
+
+require 'avalon/variations_mapping_service'
+
+# A tool for converting a variations v2p file to an Avalon timeline
+# @since 5.1.0
+module Avalon
+  # Class for the conversion of a v2p file to an Avalon timeline
+  class VariationsTimelineImporter
+    DEFAULT_TIMELINE_TITLE = 'Imported Variations Timeline'
+    @timepoint_index = 0
+
+    # Same as build_timeline but will attempt to save the timeline, all timeline items, clips, and markers
+    def import_timeline(timeline_file, user)
+      timeline_xml = parse_timeline(timeline_file)
+      build_timeline(timeline_xml, user)
+    end
+
+    private
+
+      # Takes a timeline and parses it using Nokogiri
+      # @param [File] timeline The timeline xml file
+      # @raise [ArgumentError] When the timeline cannot be parsed or is not a timeline file
+      def parse_timeline(file)
+        xml = nil
+        begin
+          xml = Nokogiri::XML(file, &:strict)
+        rescue
+          raise ArgumentError, 'File is not valid XML.'
+        end
+        raise ArgumentError, 'File is not a timeline' unless xml.xpath('//Timeline').first.present?
+        xml
+      end
+
+      def build_timeline(timeline_xml, user)
+        @timepoints ||= extract_timepoints(timeline_xml)
+
+        timeline = initialize_timeline(timeline_xml, user)
+
+        container = timeline_xml.xpath('//Timeline/@mediaContent').text
+        media_offset = timeline_xml.xpath('//Timeline/@mediaOffset').text.to_i
+        media_length = timeline_xml.xpath('//Timeline/@mediaLength').text.to_i
+
+        mf = VariationsMappingService.new.find_offset_master_file(container, media_offset)
+
+        starttime = media_offset / 1000.0
+        endtime = (media_offset + media_length) / 1000.0
+        timeline.source = Rails.application.routes.url_helpers.master_file_url(mf) + "?t=#{starttime},#{endtime}"
```

Comment:
Not sure if it is worth changing, but I'm pretty sure that the route helper can take the ID instead of the object.  If so, the Fedora load that happens in `find_offset_master_file` could be avoided and just the ID returned.